### ISSUE TITLE
Useful UI features from Harlequin missing

### ISSUE THREAD

**Issue by digitallysane** [opentoonz/opentoonz#108]
As seen here: https://youtu.be/-jjeh1USIlY?t=4m14s
the XSheet in Harlequin features display of keyframes and their interpolation and cycling. This is missing from OpenToonz.

Here: https://youtu.be/-jjeh1USIlY?t=4m37s
The chooser for the function curves is placed to the left. Also, the function curve editor is dockable.

The Edit Tool in Harlequin features a multi-purpose handle, which can move, rotate and scale depending of what part of it is used. This is also missing.

I guess all those are already in the code but must be activated in some way?

**Comment by blurymind** [comment]
Nicely spotted!
Other things I noticed that they removed:
- Ability to edit rooms in the gui - create new room, delete existing room
- Some missing key bindings: https://github.com/opentoonz/opentoonz/issues/56
- The room layouts are a mess here, they were ok in harlequin: https://github.com/opentoonz/opentoonz/issues/80

**Comment by odysseywestra** [comment]
Those are probably missing because opentoonz is based on the one that was customized by Studio Ghibli. So they were probably added afterwards or removed since they were not needed at the time.

**Comment by digitallysane** [comment]
The keyframes in the XSheet was there since forever I think (at least it was there in Toonz 5).
It's also the most useful of the lot, working blind in the XSheet and *having* to have the Functions window open to check the keyframes is not the best way.
Also the popup non dockable Curves window is usable, but not the most convenient. I'd prefer the Toonz way, where I think you can simply switch between curves and spreadsheet. And of course, left side channel chooser like any software.

**Comment by blurymind** [comment]
They definitely removed some stuff on purpose - removing the delete command key binding for example makes no sense :D

**Comment by digitallysane** [comment]
![xsheet keys](https://cloud.githubusercontent.com/assets/17073032/14121311/8a5cf154-f5ff-11e5-9931-975db0b1c196.jpg)
From the Harlequin manual.

**Comment by brambox** [comment]
Sadly OpenToonz at this stage is far from what the real Toonz is...
Right now its slow crashing often no reason things broken all over... I test it on osx windows 7 8 10 and it has issues on all.
The ui is reworked ugly it dont even have tools on by default? The mising keyframe in xsheet is  pure nightmare..
On other side Toonz Harlequin is super fast and stable and quite easy to find things.
Really I don't know what they did to it to break it so badly. 
Still having this open source is a dream come true and with time most things will be helpfully be fixed.

**Comment by blurymind** [comment]
I just found out a big one!
The Stage schematic does not allow the user to set outputs on nodes.
This causes this huge limitation: https://github.com/opentoonz/opentoonz/issues/102

In toonz harlequin you can specify outputs of nodes and as a result you get access to some of that node's data (output is called by index).
![capture](https://cloud.githubusercontent.com/assets/6495061/14124263/7ef82d5e-f5fc-11e5-99b2-e9f7fb5c5b41.PNG)
You drag on it to switch it.

**Comment by AyameLee** [comment]
Also missing was the ability to control Onion Skin settings directly in the X-Sheet as can be seen in this Harlequin instruction video right at the beginning.

https://www.youtube.com/watch?v=GigcJytbDZM

**Comment by blurymind** [comment]
Adding to the list of harlequin things opentoonz lost: https://github.com/opentoonz/opentoonz/issues/110
half the brush features for example :)

**Comment by cryovat** [comment]
It would be very interesting to know how old the Ghibli version is compared to Harlequin 7 and how much the source trees differ.

**Comment by blurymind** [comment]
A lot of this stuff looks intentionally removed. Why would they have all these style tabs, but not be able to load any of their resources?
Why would you have the functionality to parent a level to a mesh bone ,but not be able to do it because of a missing UI.
Why  would you hide the toolbox by default and remove some of the key bindings.

I am not saying that they did it to mess with us, but they did it to mess with us :P

**Comment by AyameLee** [comment]
What's interesting is I went back through all my Studio Ghibli artbooks (I have them all) and in a few of them, especially The Wind Rises they give little tutorials of how they compose a scene using Toonz. Looking closely at the screenshots, GhibliToonz _(it has a Totoro icon)_ looks even more stripped down than OpenToonz, they usually only have 1 or 2 rooms in the screenshots.

It would seem that like Retas Studio comes in four separate parts, it's by design to prevent users from being exposed to too many menus and only have access to what is essential. Since Studio Ghibli did all their frames traditionally they probably had no need for brush styles and disabled it.

However, in one of the books they talk about how they tried to create a brush that mimicked their traditional line. Perhaps that is what the default brush is in GhibliToonz/OpenToonz and the only brush users were allowed access to at Studio Ghibli?

Their traditional line is quite opaque, so they wouldn't need texture to help create the illusion. I think they used Mitsubishi pencils since Spirited Away which have a smooth quality to them when combined with animation paper, no paper grain etc.

![sg_cap_git](https://cloud.githubusercontent.com/assets/18118611/14146362/2f8c7d0e-f68f-11e5-8d0e-4014ef184b67.JPG)

**Comment by shun-iwasawa** [comment]
Hello, thank you very much for comments. 

I was in charge of development of Toonz Ghibli Edition.
OpenToonz is based on Toonz Ghibli Edition, which had been customized for production work in Studio Ghibli.
The customization was done on the basis of situations as follows:

- Each production workflow ( ink&paint, color design and composition ) is done by different staff in individual division.
- Animation was originally drawn on papers and converted to Toonz raster level in Toonz.
  (We did not draw vector-based animation, did not make cutout animation neither.)

Under the above situations, we decided to hide/remove some UI and functions from Toonz Harlequin in order to increase efficiency by limiting commands only related to each workflow.

Here are the examples:

1, We omitted a functions for create/delete rooms. Istead, we set ready-made room for each workflow which displays specialized menubar with necessary command set. (* For now, the specialized menubar is only available in Windows version.)

2, We removed buttons for moving across sub-xsheet hierarchy in the menubar. Because this UI is rarely used in ink&paint or color design work, and also because it was, in our thought, not so much inportant function as occupying large space in the menu bar area. Instead of it, a text "(Sub)" is displayed in the title bar of Xsheet panel for showing whether you are in sub or parent xsheet. Moving accross sub-xsheet is available with the cammands in context menu of Xsheet.

3, We removed "column keys"(keys displayed in xsheet columns) feature. Because setting the key on some xsheet cell causes setting keys for all of transformation parameters (n/s, e/w, scale etc.) of the column. Which was, in our thought, quite difficult to deal with. And also with this UI user may modify the positions of keys by mistake while placing levels in xsheet. So we decided to display all parameter keys only in the function editor panel.

4, Regarding the output port of stage schematic node, we did two changes for simplification:
   - Ports with letters (A,B,C...) means the amount of 8-inch-horizontal offset from the original position (with letter "B"). We removed this feature since it is less flexible than doing the same thing using pegbars.
   - Ports with numbers (1,2,3...) are for the Hook tool. We kept it alive, but very poorly...
     When you add a hook, new port will be added just below the original port with no number displayed. This is mainly because Ghibli did not use Hook tool in production and put less priority for modify it.

OpenToonz has some changes other than listed above and we suppose that "original" Toonz users may feel a confusion about it. Please let us mention again that such changes were done as the result of reflecting local inputs from Studio Ghibli and other Japanese animation productions.

For this reason, we think that just reverting the Harlequin's UI is not the best way at this time. If you would like to revert these UIs, we think there must be a option (in preferences) for adding them or not. 
We sincerely appreciate your continuous support and advice for making the better software.

Anyway, I will try to revert create/delete rooms feature and will try to modify stage schematic port for Hook tool.

Thanks.

**Comment by digitallysane** [comment]
Thanks a lot for the clarification and openness!

**Comment by blurymind** [comment]
@shun-iwasawa The changes listed make a lot of sense for a traditional workflow - which ghibly and a lot of animation studios in japan use.
Some of the changes however handicap the cut out animation style workflow - complete removing vital features that heavily rely on things that were omitted or simplified here.

The port slots redesign in the schematic view in particular has impaired the plastic tool:
please read the posts here - https://github.com/opentoonz/opentoonz/issues/102
- As a result the user can not attach levels to mesh bones created in the plastic tool's mesh node.
![meshboneparentexternalobject](https://cloud.githubusercontent.com/assets/6495061/14119745/7e910166-f5e7-11e5-9406-6cfefa029944.PNG)
![capture-outputs](https://cloud.githubusercontent.com/assets/6495061/14124184/1c467940-f5fc-11e5-8dbc-37ac63a3c048.jpg) ![missingoutputvalue](https://cloud.githubusercontent.com/assets/6495061/14124320/cafff0ce-f5fc-11e5-9424-15ef7d445b56.JPG)
This greatly limits the character rig's design.

- When using multiple hooks - these hooks have no indication in the schematic view- so it is very confusing to guess which is which:
![35d10bd2-f6ab-11e5-96b7-c428f81abf7a](https://cloud.githubusercontent.com/assets/6495061/14177741/a1ebe054-f74f-11e5-9122-684dc2e7395f.png)

This is not to criticize the workflow or design decisions as much as to request to bring back these features so open toonz is more suitable for the cutout animation style - which it was originally designed to encompass. 

Tv and low budget productions rely on it very heavily. If opentoonz does not support cutout animation style, it is going to lose a lot of the indie/small studios!

I strongly believe that having this functionality in opentoonz is much more important than how it looked visually to artists. A small toggle button to hide schematic show/view port indicators in this case would be a nice middle ground. I also proposed ways to improve the node/ports design of toonz harlequin at https://github.com/opentoonz/opentoonz/issues/102

So I beg of you and the developers of this wonderful software - please bring back this functionality in some form and make opentoonz as powerful as toonz harlequin. If you didn't like the design of some elements in harlequin - change their design - do not remove the actual functionality which that design was created to bring to the table. :)

**Comment by AyameLee** [comment]
I'm quite happy with how OpenToonz is right now since I am a frame by frame animator, I completely understand what Studio Ghibli was going for and I can immediately notice it compared to Harlequin 7.1. Functions are much faster to reach, less exposure to unnecessary objects, much slimmer UI that takes up less screen real-estate and everything else mentioned by @shun-iwasawa.

At the very least, undoing the changes made by Studio Ghibli should be user optional. Perhaps it could work like a workspace scheme seen in Clip Studio Paint, Photoshop, etc.

It's the first time in forever frame by frame animators received anything like this, I'd be disappointed if people once again focused on Flash compatibility, auto-tweening and rigging, software and plugins of which come a dime a dozen.

**Comment by blurymind** [comment]
@AyameLee > That is perfectly fine.
Yes, being able to tweak the software to offer a slimmed down interface for 2d animators is not a bad thing.

Different people (and studios) however like to work different ways. Workflow greatly impacts budget of production. Greenlighting a production depends on budget. Western and eastern tv animation production differ. Western studios tend to favor the cut out style for tv shows.

I am not telling developers to favor one workflow over another or imposing a preference or saying be like flash. I am simply requesting that features are not entirely removed or handicapped on purpose in order to favor one workflow over the other!
This is so to say that different types of animators should be treated equally.

That said opentoonz should have as a goal both the traditional and the more digital artists. Inviting all types of animators to join in and use it's wonderful tools. That way it's user and developer base will be healthier.

If you actually look at the majority of feedback when openToonz initially came out (at the english forum and on the internet as a whole)- you will find that most of people found it's slimmed down interface incredibly confusing - because it was hiding the toolbox (how can we explore what the software can do, when we cant find the tools to test?). Those people are the ones that are used to and expect to do paperless type animation in opentoonz. They couldnt find their tools, the slimming of the interface imposed tremendous usability problems on their workflow.
https://vimeo.com/160517233
The first thing they did was to change the interface in order to meet their workflow. This is where the slimming down started getting in the way - because it actually took away customization of the software. In order to customize rooms - the user now has to edit text files. The edit room menus were hidden in a right click menu - they werent taking any screen estate.

It is ok to slim down the interface, but it is not in the spirit of open source technology to impose limitations on it's users. On the contrary - opentoonz should keep all the features that toonz had and let the users decide how it's interface looks. If they want slimmed down - its a preset for them. If they want the full cutout keyframing one - they can also do that. In the end they can customize it.

My request is to fix what was broken by the slimming down - broken features such as the schematic view plastic tool missing outputs for example. And yes - if you so insist - make it all optionally visible.
In the x-sheet - bring back keyframes - but make them optionally visible. In the schematic view - bring back the output ports and their indicators, the ability to add output ports - but make them optionally visible.

The key word here is optional, not removed on purpose to suit the needs of one studio or a particular individual. :)

**Comment by AyameLee** [comment]
@blurymind I know you're not insisting developers go one route over another but many people across social media and the Google Groups board are, so it is a little worrying that there is this momentum building against what Studio Ghibli did to Toonz and the lack of understanding of it.

Perhaps not everyone has heard of Studio Ghibli but to those that have surely it must be apparent what was going to be gained and what was going to be lost.

There is little information on Toonz Premium but it appears to be identical to Harlequin 7.1. The question is why would they not just release that instead of GhibliToonz? That's why it seems redundant to me to put back what Studio Ghibli took out.

[http://www.toonzpremium.com/#!toonz/gu6j1](url)

> Toonz Premium is the meeting point of the traditional and paperless animation production workflows.
It provides state-of-the-art tools for both the approaches for endless creative possibilities, and allows you to follow your personal approach to computer-assisted animation fast-forwarding tedious and time-consuming steps.
>
>![tp1](https://cloud.githubusercontent.com/assets/18118611/14188668/b2971b94-f781-11e5-9cae-7254f1c53422.JPG)

Not to mention Digital Video appeared to be reluctant to support raster and frame by frame based animation if it wasn't for Studio Ghibli. I hope the community continues to support it.

**Comment by AyameLee** [comment]
@blurymind I know you're not insisting developers go one route over another but many people across social media and the Google Groups board are, so it is a little worrying that there is this momentum building against what Studio Ghibli did to Toonz and the lack of understanding of it.

Perhaps not everyone has heard of Studio Ghibli but to those that have surely it must be apparent what was going to be gained and what was going to be lost.

It would be great to see everything from Harlequin 7.1 put into OpenToonz, but then why not just release Toonz Premium instead? It looks identical to Harlequin 7.1. That's why it seems redundant to me to put back what Studio Ghibli took out, however many versions back.

(http://www.toonzpremium.com/#!toonz/gu6j1)

>Toonz Premium is the meeting point of the traditional and paperless animation production workflows.
 It provides state-of-the-art tools for both the approaches for endless creative possibilities, and allows you to follow your personal approach to computer-assisted animation fast-forwarding tedious and time-consuming steps.
>
>![tp1](https://cloud.githubusercontent.com/assets/18118611/14189052/438e9f18-f783-11e5-93e7-55e351ea38e2.JPG)

Not to mention Digital Video appeared to be reluctant to support raster if it wasn't for Studio Ghibli. I hope the community continues to support it.

**Comment by blurymind** [comment]
@AyameLee I love Ghibli as much as you do. I know that it seems intimidating, but you shouldn't worry that what people want to get back from toonz in opentoonz would in any way compromise it's design.
Bringing back features that were removed doesn't mean that the interface will have to be more cluttered or that you would lose any of what it currently has.
 I mentioned that we can just have them hidden - didnt I? Doesn't it make more sense to have the features, but leave it to the users which ones they use and how their interface looks like? 

I am arguing against removing features - so that more people have use of the software. Arguing against having features that you personally have no use of because you don't like their interface doesnt make any sense to me. You gain literally nothing ,but disappointing those who do need them.

Having worked as a cleanup artist and inbetweener on a number of traditional animation productions, and also as an animator with 2d character rigs - my opinion is that in order to meet the requirements of more studios - the software needs to do the best of both worlds.
That way there is more of a chance that the next production that has work for me is done with opentoonz rather than toonboom which offers the best of both worlds.

**Comment by AyameLee** [comment]
@blurymind You're probably right, this forum is much more forward looking than the other boards.

I noticed that Toonz Premium is listed as 7.2, and found this.

![tzfet](https://cloud.githubusercontent.com/assets/18118611/14193476/feef4cee-f79c-11e5-8058-608bba1c7082.JPG)

I'd love to have **measurement in pixels**!

**Comment by blurymind** [comment]
See, we both want something back from toonz harlequin before the slimming down :)
It's better to have it available and be able to hide it, instead of entirely removing it from the software.

**Comment by odysseywestra** [comment]
@digitallysane , @blurymind: Could you move this conversion over to the Google Groups Forum? We need to keep this issue tracker focused on bugs and issues. The Forum has been a organized so it shouldn't be as a mess as it was before.

**Comment by blurymind** [comment]
@odysseywestra > Yes of course. I dont think we will continue it. :) 
Applogies and thank you for organizing the forum!

Please keep this issue open until tackled. Let me know if you need any more information or assistance.

**Comment by blurymind** [comment]
@shun-iwasawa Does this fix both how hooks are displayed and also the missing mesh node outputs?
If so, you have my gratitude! Thank you!

 I made a mockup suggestion of how the schematic view output slots can be made more intuitive:
![mockup-schematic-mesh](https://cloud.githubusercontent.com/assets/6495061/14235828/0550550e-fa01-11e5-9227-c091b15d03c2.png)

**Comment by shun-iwasawa** [comment]
@blurymind My fix is just add toggle button to swap display mode of output port.
It's simpler way than you suggested in mockup, let's keep it for future issue.

See now you can grab an apple! :-)
![revert_harlequin_outputport](https://cloud.githubusercontent.com/assets/17974955/14245297/60e48986-fa9b-11e5-8434-0f1dec4b3e45.png)

**Comment by blurymind** [comment]
@shun-iwasawa This is amazing! It will make a lot of people from the opentoonz community forum very  very happy!
Thank you so much for bringing the slots back! 

And also, I love the apple grabbing baby! :+1:

**Comment by blurymind** [comment]
I closed the bug  here: https://github.com/opentoonz/opentoonz/issues/102
Also reported the fix to the opentoonz english forum, so they stop complaining about it.

**Comment by psychoanima** [comment]
WOW, thank you very much!!!

Now stupid question: How to run this fix on our local computers?

**Comment by blurymind** [comment]
@psychoanima we will have to wait for dwango to create a new build or alternatively compile one ourselves.
I am going to start compiling builds myself when the linux version becomes more stable, as compiling on linux is much easier than windows.

Btw I would like to take the opportunity to mention another cut out workflow animation style feature that has been lost from toonz harlequin:
https://github.com/opentoonz/opentoonz/issues/157
This time it's one that affects workflow speed. The Edit tool's manipulator widget has been split into modes. In old harlequin you didnt have to switch between modes to operate - everything was available in a single mode. Some people at toonz forum complained that it is a very slow way to work and I have to agree that the old way was much better. I am not saying to remove the current design in place of the old one - but to bring the old one back as a new mode called 'ALL'

**Comment by blurymind** [comment]
@shun-iwasawa  Opentoonz has some very serious regressions in the save file department.
These bugs render it quite unusable for anything but small tests. In this case it's not missing features, but broken fundamental features we have found!

These were not broken in toonz harlequin:

1. Saving a scene does not save any of the levels that it contains (https://github.com/opentoonz/opentoonz/issues/123) . The user is forced to save each individual level manually. The user is not warned that these levels will be lost when closing the application - but it warns about losing unsaved palettes. In harlequin when you save a scene - all of it's levels are also saved. So ass literally any other 2d animation software out there.

2. Saving mesh levels is completely broken. The save file dialogue would not come if you are in the plastic tool (https://github.com/opentoonz/opentoonz/issues/124). If you select another tool you do get the file save dialog, but when you do save the mesh level, opentoonz crashes completely (https://github.com/opentoonz/opentoonz/issues/198)

Somebody has a patch that adds a save all command:
https://github.com/opentoonz/opentoonz/pull/153
I am not sure if it solves above issues.

This is a critical problem. Without being able to save your progress reliably, there is no point in using the software seriously.

**Comment by blurymind** [comment]
Here is a regression bug in the mesh drawing in the viewport. In opentoonz the mesh has a black background instead of alpha. Toonz harlequin properly displays the alpha channel.

I explained how it can hinder workflow in my bug report #134 

This is in toonz harlequin:
![blackbgregressionmesh](https://cloud.githubusercontent.com/assets/6495061/14704927/124d2dd4-07af-11e6-8955-a9602e633c3c.JPG)

This is opentoonz
![opentoonz-glithch](https://cloud.githubusercontent.com/assets/6495061/14235397/2835cdee-f9f5-11e5-8d11-f5dfe78f217b.JPG)

**Comment by shun-iwasawa** [comment]
Please let me close this issue, since most of problems mentioned in this issue are already posted as other individual issues as follows.
- Ability to control Onion Skin settings directly in XSheet is missing #231
- Saving a scene does not save any of the levels #123 ("Save All" command is added in the pull request https://github.com/opentoonz/opentoonz/pull/230 for this issue)
- The Edit tool's manipulator widget #157 
- Function Curve needs to press Ctrl key for editing #159 
- Mesh has a black background instead of alpha #134

**Comment by blurymind** [comment]
@shun-iwasawa Thank you for bringing a lot of these features back! Your contributions to opentoonz have made the application much more usable than it's initial release. I look forward to the improvements you make to it every week.

 Yes, I decided to post them as individual issues in the tracker - so as to be able to collect data from more users who have experienced them. But each time I found out that they are regressions (missing or broken as compared to harlequin), I decided that it would help you identify and tackle them if I refer to them in this issue thread. 

Appologies if this has created more work for you. I will close all the issues you tackled to.
